### PR TITLE
nfs: fix nfs if multus is enabled

### PR DESCRIPTION
**Pull request by parth-gr**
Closes: https://github.com/rook/rook/issues/10812
Signed-off-by: parth-gr <paarora@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by parth-gr** [comment]
```
2022-09-12 14:28:08.176956 E | ceph-block-pool-controller: failed to reconcile CephBlockPool "rook-ceph/builtin-nfs1". failed to create pool "builtin-nfs1".: failed to create pool "builtin-nfs1".: failed to initialize pool ".nfs". : rbd: pool already registered to a different application.
2022-09-12 14:28:18.783975 I | ceph-block-pool-controller: creating pool ".nfs" in namespace "rook-ceph"
2022-09-12 14:28:18.784005 D | exec: Running command: ceph osd crush rule create-replicated .nfs default osd --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-12 14:28:19.140043 D | exec: Running command: ceph osd pool get .nfs all --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-12 14:28:19.500864 D | exec: Running command: ceph osd pool application get .nfs --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-12 14:28:19.887752 I | cephclient: application "nfs" is already set on pool ".nfs"
2022-09-12 14:28:19.887774 I | cephclient: reconciling replicated pool .nfs succeeded
2022-09-12 14:28:19.887779 D | cephclient: checking that pool ".nfs" has the failure domain "osd"
2022-09-12 14:28:19.887793 D | exec: Running command: ceph osd pool get .nfs all --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-12 14:28:20.348578 D | exec: Running command: ceph osd crush rule dump .nfs --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-12 14:28:20.708101 D | cephclient: pool ".nfs" has the expected failure domain "osd"
2022-09-12 14:28:20.708122 I | ceph-block-pool-controller: initializing pool ".nfs"
2022-09-12 14:28:20.714468 D | exec: ExecWithOptions {Command:[timeout 15 rbd pool init .nfs] Namespace:rook-ceph PodName:rook-ceph-mgr-a-575dc9b4b-s6vlg ContainerName:cmd-proxy Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:false}
2022-09-12 14:28:20.841459 D | ceph-block-pool-controller: pool "rook-ceph/builtin-nfs1" status updated to "Failure"
2022-09-12 14:28:20.841501 E | ceph-block-pool-controller: failed to reconcile CephBlockPool "rook-ceph/builtin-nfs1". failed to create pool "builtin-nfs1".: failed to create pool "builtin-nfs1".: failed to initialize pool ".nfs". : rbd: pool already registered to a different application.
```

**Comment by parth-gr** [comment]
@Rakshith-R 

```
2022-09-13 10:24:28.555366 D | ceph-nfs-controller: successfully configured CephNFS "rook-ceph/my-nfs"
2022-09-13 10:24:28.575992 D | ceph-block-pool-controller: successfully configured CephBlockPool "rook-ceph/builtin-nfs"
2022-09-13 10:24:38.555891 D | ceph-spec: "ceph-nfs-controller": CephCluster resource "my-cluster" found in namespace "rook-ceph"
2022-09-13 10:24:38.556442 D | ceph-spec: "ceph-nfs-controller": ceph status is "HEALTH_WARN", operator is ready to run ceph command, reconciling
2022-09-13 10:24:38.602382 D | op-k8sutil: ConfigMap ceph-nfs-controller-detect-version is already deleted
2022-09-13 10:24:41.578673 I | ceph-block-pool-controller: creating pool ".nfs" in namespace "rook-ceph"
2022-09-13 10:24:41.578690 D | exec: Running command: ceph osd crush rule create-replicated .nfs default osd --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-13 10:24:45.877435 D | exec: Running command: ceph osd pool get .nfs all --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-13 10:24:47.372830 D | exec: Running command: ceph osd pool create .nfs 0 replicated .nfs --size 1 --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-13 10:24:49.952052 D | CmdReporter: job ceph-nfs-controller-detect-version has returned results
2022-09-13 10:24:49.999621 D | ceph-cluster-controller: hot-plug cm watcher: only reconcile on hot plug cm changes, this "ceph-nfs-controller-detect-version" cm is handled by another watcher
2022-09-13 10:24:50.100607 D | ceph-spec: object "ceph-nfs-controller-detect-version" did not match on delete
2022-09-13 10:24:50.100638 D | ceph-spec: object "ceph-nfs-controller-detect-version" did not match on delete
2022-09-13 10:24:50.100660 D | ceph-spec: object "ceph-nfs-controller-detect-version" did not match on delete
2022-09-13 10:24:50.100667 D | ceph-spec: object "ceph-nfs-controller-detect-version" did not match on delete
2022-09-13 10:24:53.162408 D | exec: Running command: ceph osd pool application get .nfs --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-13 10:24:53.252543 I | ceph-nfs-controller: configuring pool ".nfs" for nfs
2022-09-13 10:24:53.252597 D | exec: Running command: ceph osd pool create .nfs --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-13 10:24:56.282725 D | exec: Running command: ceph osd pool application enable .nfs nfs --yes-i-really-mean-it --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-13 10:24:56.328831 D | exec: Running command: ceph osd pool application enable .nfs nfs --yes-i-really-mean-it --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-13 10:25:00.387387 I | cephclient: reconciling replicated pool .nfs succeeded
2022-09-13 10:25:00.387416 D | cephclient: checking that pool ".nfs" has the failure domain "osd"
2022-09-13 10:25:00.387432 D | exec: Running command: ceph osd pool get .nfs all --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-13 10:25:00.387657 I | ceph-nfs-controller: set pool ".nfs" for the application nfs
2022-09-13 10:25:00.387663 D | ceph-nfs-controller: reconciling ceph nfs deployments
2022-09-13 10:25:00.396887 I | ceph-nfs-controller: updating ceph nfs "my-nfs"
2022-09-13 10:25:00.405698 D | exec: Running command: rados --pool .nfs --namespace my-nfs --conf /var/lib/rook/rook-ceph/rook-ceph.config stat conf-nfs.my-nfs
2022-09-13 10:25:03.028596 D | exec: Running command: ceph osd crush rule dump .nfs --connect-timeout=15 --cluster=rook-ceph --conf=/var/lib/rook/rook-ceph/rook-ceph.config --name=client.admin --keyring=/var/lib/rook/rook-ceph/client.admin.keyring --format json
2022-09-13 10:25:04.732447 D | cephclient: pool ".nfs" has the expected failure domain "osd"
2022-09-13 10:25:05.971411 D | ceph-spec: update event from a CR: "builtin-nfs"
2022-09-13 10:25:05.995756 D | ceph-block-pool-controller: pool "rook-ceph/builtin-nfs" status updated to "Ready"
2022-09-13 10:25:05.995801 D | ceph-block-pool-controller: successfully configured CephBlockPool "rook-ceph/builtin-nfs"
```

THe log looks fine to me, still nfs pod not coming

File: pkg/operator/ceph/client/controller.go [hunk 1]
```diff
@@ -176,7 +176,7 @@ func (r *ReconcileCephClient) reconcile(request reconcile.Request) (reconcile.Re
 	}
 
 	// Populate clusterInfo during each reconcile
-	r.clusterInfo, _, _, err = opcontroller.LoadClusterInfo(r.context, r.opManagerContext, request.NamespacedName.Namespace)
+	r.clusterInfo, _, _, err = opcontroller.LoadClusterInfo(&cephCluster.Spec, r.context, r.opManagerContext, request.NamespacedName.Namespace)
```
Comment (subhamkrai) [inline pkg/operator/ceph/client/controller.go:179]:
```suggestion
	r.clusterInfo, _, _, err = opcontroller.LoadClusterInfo( r.context, r.opManagerContext, request.NamespacedName.Namespace,&cephCluster.Spec)
```
generally we keep `context` and `namespace` before other args

**Review by subhamkrai — CHANGES_REQUESTED** [review]
similar in other places.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
This looks ready to go minus a few small things. Thanks for looking into this :)

File: pkg/operator/ceph/controller/cluster_info.go [hunk 2]
```diff
@@ -148,6 +148,11 @@ func CreateOrLoadClusterInfo(clusterdContext *clusterd.Context, context context.
 		return nil, maxMonID, monMapping, errors.Wrap(err, "failed to get mon config")
 	}
 
+	// update clusterInfo if Multus is enabled
+	if cephClusterSpec.Network.IsMultus() {
+		clusterInfo.NetworkSpec.Provider = "multus"
+	}
```
[the remark this answers is not in the document]
Comment (parth-gr) [reply pkg/operator/ceph/controller/cluster_info.go]: okay

**Comment by parth-gr** [comment]
This would need a follow up investigation for 
`rook-ceph-nfs-my-nfs-a-5cfb7f5978-zsjsg        1/2     CrashLoopBackOff   1 (5s ago)   14s` 
As the nfs pod that is started is in CLBO state

File: .github/workflows/canary-integration-test.yml [hunk 3]
```diff
@@ -1421,7 +1424,7 @@ jobs:
       - name: set up tmate session for debugging
         if: failure() && env.USE_TMATE
         uses: mxschmitt/action-tmate@v3
-        timeout-minutes: 60
+        timeout-minutes: 600
```
Comment (subhamkrai) [inline .github/workflows/canary-integration-test.yml]: revert this since not required.

**Review by subhamkrai — CHANGES_REQUESTED** [review]
change the pr and commit title to `nfs`

File: .github/workflows/canary-integration-test.yml (hunk 3 above)
Comment (parth-gr) [reply .github/workflows/canary-integration-test.yml]: So it is been reverted in the second commit

**Review by BlaineEXE — APPROVED** [review]

**Review by subhamkrai — APPROVED** [review]

File: .github/workflows/canary-integration-test.yml (hunk 3 above)
Comment (parth-gr) [reply .github/workflows/canary-integration-test.yml]: Yes right

**Pull request merged**